extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -89,29 +229,36 @@ public Signature boundSignature() {
     public Object execute(int idxInPartition,
                           WindowFrameState currentFrame,
                           List<? extends CollectExpression<Row, ?>> expressions,
-                          Input... args) {
+                          boolean ignoreNulls,
+                          Input[] args) {
         final int offset;
         if (args.length > 1) {
             Object offsetValue = args[1].value();
-            if (offsetValue == null) {
-                return null;
-            } else {
+            if (offsetValue != null) {
                 offset = ((Number) offsetValue).intValue();
+            } else {
+                return null;
             }
         } else {
             offset = 1;
         }
-
-        var lagRowCells = currentFrame
-            .getRowInPartitionAtIndexOrNull(offsetDirection.getTargetIndex(idxInPartition, offset));
-        if (lagRowCells != null) {
-            var lagRow = new RowN(lagRowCells);
-            for (CollectExpression<Row, ?> expression : expressions) {
-                expression.setNextRow(lagRow);
-            }
-
-            return args[0].value();
-        } else {
+        if (offset == 0) {
+            ignoreNulls = false;
+        }
+        // if offset is not constant but varying per iteration, OffsetDirection.cachedNonNullIndex cannot be utilized
+        if (idxInPartition == 0 || (cachedOffset != null && cachedOffset != offset)) {
+            offsetDirection = resolveOffsetDirection(offset);
+        }
```

Comment:
Might be safer to check if the `Input` is a `Literal` to see if it is static or not.  I think this logic here may be error prone if it appears as if the offset is static because of successive equal values.